javav2/example_code/redshift/src/main/java/com/example/redshift/scenario/RedshiftActions.java

```diff
@@ -0,0 +1,506 @@
+    /**
+     * Checks the status of an SQL statement asynchronously and handles the completion of the statement.
+     *
+     * @param sqlId the ID of the SQL statement to check
+     * @return a {@link CompletableFuture} that completes when the SQL statement's status is either "FINISHED" or "FAILED"
+     */
+    public CompletableFuture<Void> checkStatementAsync(String sqlId) {
+        DescribeStatementRequest statementRequest = DescribeStatementRequest.builder()
+            .id(sqlId)
+            .build();
+
+        return getAsyncDataClient().describeStatement(statementRequest)
+            .thenCompose(response -> {
+                String status = response.statusAsString();
+                logger.info("... Status: {} ", status);
+
+                if ("FAILED".equals(status)) {
+                    throw new RuntimeException("The Query Failed. Ending program");
+                } else if ("FINISHED".equals(status)) {
+                    return CompletableFuture.completedFuture(null);
+                } else {
+                    // Sleep for 1 second and recheck status
+                    return CompletableFuture.runAsync(() -> {
+                        try {
+                            TimeUnit.SECONDS.sleep(1);
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException("Error during sleep: " + e.getMessage(), e);
+                        }
+                    }).thenCompose(ignore -> checkStatementAsync(sqlId)); // Recursively call until status is FINISHED or FAILED
+                }
+            }).whenComplete((result, exception) -> {
+                if (exception != null) {
+                    // Handle exceptions
+                    logger.info("Error: {} ", exception.getMessage());
+                } else {
+                    logger.info("The statement is finished!");
+                }
+            });
+    }
+    // snippet-end:[redshiftdata.java2.checkstatement.main]
+
+    // snippet-start:[redshiftdata.java2.getresults.main]
+    /**
+     * Asynchronously retrieves the results of a statement execution.
+     *
+     * @param statementId the ID of the statement for which to retrieve the results
+     * @return a {@link CompletableFuture} that completes when the statement result has been processed
+     */
+    public CompletableFuture<Void> getResultsAsync(String statementId) {
+        GetStatementResultRequest resultRequest = GetStatementResultRequest.builder()
+            .id(statementId)
+            .build();
+
+        return getAsyncDataClient().getStatementResult(resultRequest)
+            .handle((response, exception) -> {
+                if (exception != null) {
+                    logger.info("Error getting statement result {} ", exception.getMessage());
+                    throw new RuntimeException("Error getting statement result: " + exception.getMessage(), exception);
+                }
```

Comment:
Can we do any better with the error handling here? Any specific error codes that a user might encounter related to this service in this scenario?

Reply:
For Async - this is wrapped in a Runtime and thrown back to calling code. Callign code determines the issue and displays issue.